### PR TITLE
ESQL planning and streaming changes for bulk keyword lookup

### PR DESCRIPTION
**Pull request by cimequinox**
<table>
    <tbody>
        <tr>
            <td>ES|QL LOOKUP JOIN command</td>
            <td><a href="https://www.elastic.co/docs/reference/query-languages/esql/commands/lookup-join">https://www.elastic.co/docs/reference/query-languages/esql/commands/lookup-join</a></td>
        </tr>
        <tr>
            <td>Related&nbsp;pull&nbsp;requests</td>
            <td><a href="https://github.com/elastic/elasticsearch/pull/144704">144704</a> ES|QL Improve LOOKUP JOIN on single keyword</td>
        </tr>
        <tr>
            <td rowspan=3>Related issues</td>
            <td><a href="https://github.com/elastic/elasticsearch/issues/147030">147030</a> <b>ESQL planning and streaming changes for bulk keyword lookup</b> </td>
        </tr>
        </tr>
            <td><a href="https://github.com/elastic/elasticsearch/pull/144760">144760</a> ESQL: Expression-based LOOKUP JOIN extracts right join key</td>
        </tr>
        <tr>
            <td><a href="https://github.com/elastic/elasticsearch/pull/141109">141109</a> Optimization when result is small after filter push-down on right</td>
        </tr>
    </tbody>
</table>

# Background

[ESQL 144704][] added the bulk keyword lookup optimization to the pre-9.4 non-streaming lookup join path.

This PR adds the optimization to the streaming path using the local lookup join planning logic added in 9.4.

This is a local physical optimization planned on the lookup node.

Most work here is done in the `LookupPhysicalPlanOptimizer` and `LookupExecutionPlanner`.

# Design

## lookup execution planning

[ESQL 147030][] extends local execution planing as follows.

### ParameterizedQueryExec

Additional `bulkLookupLeft` and `bulkLookupRight` members are added to `ParameterizedQueryExec` in the physical plan.

These attributes default to `null` and are not serialized.

A new `LuceneBulkLookup` rule in the `LookupPhysicalPlanOptimizer` sets these members 
to the left and right attributes of the join when the bulk lookup optimization can be taken.

#### example query and portion of physical plan

```
ROW name = "French"
| LOOKUP JOIN languages_lookup ON name == language_name

...
  \_ ParameterizedQueryExec[
     matchFields     = [ name ],
     joinOn          = name == language_name,
     output          = [ _doc, _positions ],
     bulkLookupLeft  = name, ◀────────────────── indicates bulk lookup will be used
     bulkLookupRight = language_name ◀────────── indicates bulk lookup will be used
   ]
```

### BulkLookupMvFilterExec

This physical plan node removes false-positive multivalue matches returned from the bulk keyword optimization.

It is present only on the lookup node and is not serialized.

#### example query and portion of physical plan

```
ROW name = "French"
| LOOKUP JOIN languages_lookup ON name == language_name

...
  \_ BulkLookupMvFilterExec[ language_name{f}#2 ] ◀─ handles false-positive multivalues in right join field
    \_ ParameterizedQueryExec[
       matchFields     = [ name ],
       joinOn          = name == language_name,
       output          = [ _doc, _positions ],
       bulkLookupLeft  = name{r}#4,
       bulkLookupRight = language_name{f}#2
     ]
```

### LuceneBulkLookup

During local physical optimization on the lookup node, `LookupPhysicalPlanOptimizer` uses
this rule to examine the join expression within the `ParameterizedQuery`.

The rule sets `bulkLookupLeft` and `bulkLookupRight` to the respective attributes in the join 
when the expression meets the equivalent conditions to those checked by `ExpressionQueryList` 
in [ESQL 144704][] on the non-streaming path:

* no `AliasFilter` is present
* no Lucene query is present
* the join contains a single **EsqlBinaryComparison** with function type **BinaryComparisonOperation.EQ**
* the left and right expressions of the comparisons are Attributes
* the left and right data types are KEYWORD

The rule handles 
* `lookup join on <keyword-field>` 
* `lookup join on <left-keyword-field> = <right-keyword-field>` 

When these conditions hold, `LuceneBulkLookup` replaces the original `ParameterizedQueryExec` 
with a `BulkLookupMvFilterExec` containing a `FieldExtractExec` with a `ParameterizedQueryExec` 
whose `bulkLookupLeft` and `bulkLookupRight` members are set to the respective attributes in the join.

#### example update to the physical plan

```
...
  \_ BulkLookupMvFilterExec[ language_name{f}#2 ] ◀─ added by LuceneBulkLookup
    \_ ParameterizedQueryExec[                  ◀─── added by LuceneBulkLookup
       matchFields     = [ name ],
       joinOn          = name == language_name,
       output          = [ _doc, _positions ],
       bulkLookupLeft  = name{r}#4,
       bulkLookupRight = language_name{f}#2
     ]
```

### LookupPhysicalPlanOptimizerContext

A new subclass of `LocalPhysicalOptimizerContext` called `LookupPhysicalPlanOptimizerContext` 
is used to hold the `AliasFilter` needed by the `LuceneBulkLookup` optimizer rule.

### LookupPhysicalPlanOptimizer

This class's `RULES` hold the additional `LuceneBulkLookup` rule, which relies on
the existing `InsertFieldExtraction` rule to insert the `FieldExtractExec` 
needed by `BulkLookupMvFilterExec`.

The `LuceneBulkLookup` rule runs before `ReplaceSourceAttributes` so that it may
identify the right join field in the ParameterizedQueryExec plan output.

#### example query and final form of the physical plan

```
ROW name = "French"
| LOOKUP JOIN languages_lookup ON name == language_name

ProjectExec[ $$Positions$${r}#0, language_code{f}#1, language_name{f}#2 ]
\_ FieldExtractExec[ language_code{f}#1 ]
  \_ BulkLookupMvFilterExec[ language_name{f}#2 ]
    \_ FieldExtractExec[ language_name{f}#2 ]
      \_ ParameterizedQueryExec[
         matchFields     = [ name ],
         joinOn          = name == language_name,
         output          = [ _doc, _positions ],
         bulkLookupLeft  = name{r}#4,
         bulkLookupRight = language_name{f}#2
       ]
```

### LookupExecutionPlanner

New methods for planing

* `planBulkLookupMvFilterExec` creates the physical operation for the `BulkLookupMvFilterExec` physical plan

* `getBulkKeywordQueryGenerator` creates the `BulkKeywordQueryGenerator` for the `LookupQueryOperatorFactory` returned by `planParameterizedQueryExec`

#### LookupQueryOperatorFactory

This now uses `getBulkKeywordQueryGenerator` which returns

* a `BulkKeywordQueryGenerator` as a queryList when the optimization is used
* null otherwise

When the optimization is not in effect the factory falls back to using `queryListFromPlanFactory.create` to obtain a queryList.

### BulkLookupQueryGenerator

This is a private class which implements the portion of the `LookupEnrichQueryGenerator` interface needed by the `LookupQueryOperator`.

* `getPosition`
* `getBulkKeywordLookup`

`getQuery` throws an unimplemented exception because `LookupQueryOperator` should not call it when there's a `BulkKeywordLookup`.

### LookupQueryOperator

When no optimization is in effect, `LookupQueryOperator` uses the queryList to issue Lucene queries.

With the bulk lookup optimization, `LookupQueryOperator` does not issue Lucene queries. 
Instead it uses a more efficient Lucene api to directly accesses the docids of matching terms
as explained in [ESQL 144704][]

<br/>

# Additional Implementation Details

### LocalMapper

* Passes initial null values to `ParameterizedQueryExec` for `bulkLookupLeft` and `bulkLookupRight`

### LookupPruneFilters and ReplaceSourceAttributes

* Updates these rule to handle `bulkLookupLeft` and `bulkLookupRight`

### BulkLookupMvFilterOperatorFactory

* Returns a `FilterOperator` with a `BulkLookupSingleValued` when operators are constructed.

<br/>

# Tests

[ESQL 147030][] updates tests as follows.

### LookupGoldenTests

* Updates existing golden test output to include `bulkLookupLeft` and `bulkLookupRight`

### LookupPhysicalPlanOptimizerTests

* Updates tests to handle `bulkLookupLeft` and `bulkLookupRight`
* adds `testBulkLookupOnFieldOptimization` and `testBulkLookupOnExpressionOptimization` to check the optimized plan.

### LookupQueryOperatorTests

* Updates existing `LookupQueryOperator` tests and adds tests with `BulkKeywordLookup`.

[ESQL 144704]: https://github.com/elastic/elasticsearch/pull/144704
[ESQL 147030]: https://github.com/elastic/elasticsearch/issues/147030
[ESQL 144760]: https://github.com/elastic/elasticsearch/issues/144760
[ESQL 141109]: https://github.com/elastic/elasticsearch/issues/141109

## Todo

- [x] update design doc https://github.com/elastic/elasticsearch/issues/147032
- [x] update this PR description
- [x] address test failures
- [x] address draft-review feedback from Julian
   - [x] perform AliasFilter check once when rule decides if optimization is possible
     - [x] don't alter the LocalPhysicalOptimizerContext
     - [x] instead add the filter to a subclass e.g. LookupPhysicalOptimizerContext
   - [x] maintain a single instance of BulkKeyworkLookup in the BulkKeywordLookupGenerator
   - [x] update `applyBulkLookup` to handle field based join
   - [x] update `LookupPhysicalPlanOptimizerTests` for field and expression based join
   - [x] negative tests 
     - [x] aliasFilter != null && != EMPTY
     - [x] ~plan.query() != null (i.e., a Lucene query was already pushed down to the source)~ see [comment](https://github.com/elastic/elasticsearch/pull/148937#discussion_r3326421711)
     - [x] the join key data type is not KEYWORD
     - [x] the binary comparison is not EQ (e.g., <)
     - [x] ~one side of the expression is not a plain Attribute~ see [comment](https://github.com/elastic/elasticsearch/pull/148937#discussion_r3326421711)

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                    ),
+                    matchField.fieldName()
+                );
+            }
+        }
+        */
+
+        // LOOKUP JOIN ON leftKeyword == rightKeyword
+
+        if (expr instanceof EsqlBinaryComparison binaryComparison
```
Comment (julian-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: This condition is NOT in sync with the condition to apply the optimization, e.g. the Alias check. 
So we insert BulkLookupMvFilterExec but  we didnt end up taking the optimization

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java [hunk 2]
```diff
@@ -490,9 +532,62 @@ public Operator get(DriverContext driverContext) {
...
+        }
+
+        @Override
+        public int getPositionCount(Page inputPage) {
+            return inputPage.getPositionCount();
+        }
+
+        @Override
+        public BulkKeywordLookup getBulkKeywordLookup() {
+            return new BulkKeywordLookup(rightFieldName, matchChannelOffset, -1, warnings);
```
Comment (julian-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java]: LookupExecutionPlanner.java:590 creates a new BulkKeywordLookup on every call → its termsEnumCache field is null → 
  BulkKeywordLookup.java:77 reads termsEnumCache[leafOrd] → NullPointerException.

I think it will crash if the optimization is applied and we have streaming lookup join with 2+ pages.

**Review by julian-elastic — CHANGES_REQUESTED** [review]
Overall, looks good! I did a first round code review, found 2 serious problems, please address

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java (hunk 1 above)
Comment (julian-elastic) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: Think of a way to make the decision just once and affect 2 places. Or at least, both places should call the same API that decides go/no go for the optimization.

**Comment by julian-elastic** [comment]
And of course, get this to work with Field based Join, it is commented out right now

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java (hunk 1 above)
Comment (cimequinox) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: I didn't include the AliasFilter check here initially because I didn't think it was available.
I've added it to the LocalPhysicalOpimizerContext and changed the rule to do the check here.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java (hunk 2 above)
Comment (cimequinox) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java]: I didn't realize this could be called multiple times as you describe.
I will change the BulkLookupQueryGenerator to create and maintain a single instance.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java (hunk 1 above)
Comment (julian-elastic) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: I still think we are making the decision in 2 places right now and they can get out of sync. 
Here you decided to use BulkLookup. 
So you create the BulkLookupMvFilterExec. That is great. 
Now how does the ParameterizedQueryExec know about that decision? You already made the decision here, you should not recheck again. 

My concern here, is the 2 different decision points getting out of sync a few refactors down the line. We need to decide once BulkLookup or not and stick with it.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java (hunk 2 above)
Comment (julian-elastic) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/LookupExecutionPlanner.java]: Thanks for addressing this!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java (hunk 1 above)
Comment (julian-elastic) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: For the AliasFilter, I don't think you can just store it in LocalPhysicalOpimizerContext as it is now. 
LocalPhysicalOpimizerContext is shared for the LookupPhysicalOptimizer and LocalPhysicalOptimizer... 
But the LocalPhysicalOptimizer is not necessarily for single index? You need to check, but it might need to be Map<Index, AliasFilter>. It is bad design to have a single AliasFilter there, because they cannot be merged if there is more than one. 


Alternatively, maybe you can do 
```
public class LookupPhysicalOptimizerContext extends LocalPhysicalOptimizerContext {
      private final AliasFilter aliasFilter;
  }
```

**Review by julian-elastic — CHANGES_REQUESTED** [review]
I did some research for you, I think we need some more work for sharing the decision and passing the alias.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java (hunk 1 above)
Comment (cimequinox) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:116]: #### Re: decision
I think we want the decision to be made in just one place - `LuceneBulkLookup` rule.
The other place(s) should only carry out that decision.  I.e. once we've decided to take the optimization we have to do two things
- ensure the `LookupQueryOperator` has a `BulkKeywordLookup` to work with
- ensure false-positives are filtered out by the operator resulting from the `BulkLookupMvFilterExec`

`ParameterizedQueryExec` holds two members, `bulkLookupLeft` and `bulkLookupRight`, which reflect the decision.  When those two are null the optimization won't be taken.  When those are set to the left and right join attribute respectively by `LuceneBulkLookup` the `LookupQueryOperator` should use the `BulkKeywordLookup` provided to it via the `LookupQueryOperatorFactory`.

I realized there's no need for `bulkLookupLeft` and `bulkLookupRight` in `ParameterizedQuery` since they are only needed on the lookup node after local optimization, so I've removed them.   I've moved the comment on those to `ParameterizedQueryExec`.  I can add additional comments if needed to further clarify the decision details.

#### Re: AliasFilter
I didn't realize the scope of sharing of `LocalPhysicalOptimizerContext` so I've removed the `AliasFilter` I added earlier and changed the logic to use a new `LookupPhysicalOptimizerContext` with an `AliasFilter` as suggested.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import org.elasticsearch.xpack.esql.plan.physical.PhysicalPlan;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.optimizer.rules.logical.OptimizerRules.TransformDirection.UP;
+
+/**
+ * Checks {@link ParameterizedQueryExec} nodes to see if the conditions for the bulk lookup optimization are met.
+ * Sets the useBulkLookup flag when they are.
+ */
```
Comment (cimequinox) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LuceneBulkLookup.java:32]: Doh!  Missed this stale comment.  Will correct before merge.
This should read "sets bulkLookupLeft and bulkLookupRight when they are"

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/BulkLookupMvFilterExec.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.AttributeSet;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+
+import java.io.IOException;
+import java.util.Objects;
+
```
Comment (julian-elastic) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/BulkLookupMvFilterExec.java:17]: Missing Javadoc for new class

File: x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/optimizer/golden_tests/LookupGoldenTests/testConstantFieldMatchingFilter/lookup_physical_optimization.expected [hunk 5]
```diff
@@ -1,3 +1,3 @@
 ProjectExec[[$$Positions$${r}#0, language_name{f}#1]]
 \_FieldExtractExec[language_name{f}#1]<[],[],[]>
-  \_ParameterizedQueryExec[[_doc{f}#2, $$Positions$${r}#0],[MatchConfig[fieldName=language_code, channel=0, type=INTEGER]],null,null,false]
\ No newline at end of file
+  \_ParameterizedQueryExec[[_doc{f}#2, $$Positions$${r}#0],[MatchConfig[fieldName=language_code, channel=0, type=INTEGER]],null,null,false,null,null]
```
Comment (julian-elastic) [inline x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/optimizer/golden_tests/LookupGoldenTests/testConstantFieldMatchingFilter/lookup_physical_optimization.expected:3]: This looks like a bug, why are the 2 new arguments always null, null even when the optimization is applied?

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/BulkKeywordLookup.java [hunk 6]
```diff
@@ -31,6 +32,8 @@
  */
 
 public class BulkKeywordLookup {
+    private static final Logger logger = LogManager.getLogger(BulkKeywordLookup.class);
```
Comment (julian-elastic) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/BulkKeywordLookup.java:35]: Missing negative tests to confirm the optimization is not triggered when it should not be e.g. 
- aliasFilter != null && != EMPTY
- plan.query() != null (i.e., a Lucene query was already pushed down to the source)
- the join key data type is not KEYWORD
- the binary comparison is not EQ (e.g., <)
- one side of the expression is not a plain Attribute

**Review by julian-elastic — APPROVED** [review]
Thank you for driving this to completion and addressing my previous feedback! I found some nits below but they seem very easy to address. You can ask Cursor/Claude to add the missing tests for you.

File: x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/optimizer/golden_tests/LookupGoldenTests/testConstantFieldMatchingFilter/lookup_physical_optimization.expected (hunk 5 above)
Comment (cimequinox) [reply x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/optimizer/golden_tests/LookupGoldenTests/testConstantFieldMatchingFilter/lookup_physical_optimization.expected:3]: In that test language_code is an INTEGER
```
[MatchConfig[fieldName=language_code, channel=0, type=INTEGER]
```
the query is
```
FROM employees
| RENAME languages AS language_code
| LOOKUP JOIN languages_lookup ON language_code
| WHERE language_name == "English"
```
Since language_code is not a KEYWORD, the optimization isn't applied.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/BulkKeywordLookup.java (hunk 6 above)
Comment (cimequinox) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/BulkKeywordLookup.java:35]: I believe we have some incidental negative coverage e.g. the `testConstantFieldMatchingFilter` test [mentioned above](https://github.com/elastic/elasticsearch/pull/148937#discussion_r3311887397) happens to also serve as a negative test when the join key data type is not KEYWORD.  

I'll add comments to existing tests to call out negative cases already covered and add new tests for the remaining missing cases.

Comment (cimequinox) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/BulkKeywordLookup.java:35]: I've updated the tests to provide better negative coverage
- **aliasFilter != null && != EMPTY**
  Now covered by [testKeywordLookupOnFieldWithAliasFilter](https://github.com/elastic/elasticsearch/pull/148937/changes#diff-3f854c35b35318da8c7f8bf9384bf9a9a0014f08c02a9eda7284cb21f59e4961R50) and [testKeywordLookupOnExpressionWithAliasFilter](https://github.com/elastic/elasticsearch/pull/148937/changes#diff-3f854c35b35318da8c7f8bf9384bf9a9a0014f08c02a9eda7284cb21f59e4961R73)

- **plan.query() != null (i.e., a Lucene query was already pushed down to the source)**
   We cannot easily write a test for this case because with this change, the LookupPhysicalPlanOptimizer [attempts the bulk optimization before lucene pushdown](https://github.com/cimequinox/elasticsearch/blob/3a1be429e9f1d1bce004b78c9eacfda24774df8e/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LookupPhysicalPlanOptimizer.java#L40).  For instance, [testKeywordLookupWithPushableFilter](https://github.com/elastic/elasticsearch/pull/148937/changes#diff-3f854c35b35318da8c7f8bf9384bf9a9a0014f08c02a9eda7284cb21f59e4961R102) shows the bulk optimization is used instead of the lucene pushdown.

- **the join key data type is not KEYWORD**
   Covered by existing [testConstantFieldMatchingFilter](https://github.com/elastic/elasticsearch/pull/148937/changes#diff-3f854c35b35318da8c7f8bf9384bf9a9a0014f08c02a9eda7284cb21f59e4961L129)

- **the binary comparison is not EQ (e.g., <)**
   Now covered by [testKeywordLookupOnNonEqualExpression](https://github.com/elastic/elasticsearch/pull/148937/changes#diff-3f854c35b35318da8c7f8bf9384bf9a9a0014f08c02a9eda7284cb21f59e4961R85)

- **one side of the expression is not a plain Attribute**
   The [support we provide](https://www.elastic.co/docs/reference/query-languages/esql/commands/lookup-join#parameters) for lookup join expressions is limited to AND/OR combinations of simple boolean conditions of attributes.   The LogicalPlanBuilder will [generate an error](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java#L1093) when one side of an expression is not an attribute and while the left attribute may be fabricated with an EVAL, that's a supported case so we can't write a negative test for that.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/BulkLookupMvFilterExec.java (hunk 4 above)
Comment (cimequinox) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/BulkLookupMvFilterExec.java:17]: I've added a [JavaDoc for BulkLookupMvFilterExec](https://github.com/cimequinox/elasticsearch/blob/3a1be429e9f1d1bce004b78c9eacfda24774df8e/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/BulkLookupMvFilterExec.java#L18-L27)

**Comment by cimequinox** [comment]
Buildkite benchmark this with esql-joins please
